### ISSUE TITLE
Multiple Symfony Form submits fill up memory 

### ISSUE THREAD

**Issue by laurenz-glueck** [symfony/symfony#45602]
### Symfony version(s) affected

5.4.4

### Description

I have an array of items (also as array) coming form a CSV file. Now I'm iterating over the items to validate the data using Symfony forms.

What I noticed now is that with every validation the memory usage goes up by around 0.29 MB and this memory never gets released. This for example leads to a memory increase from 20 to 70 MB after validating 200 items. (Each of the items is an array with 10 key / value pairs).

The code I'm running looks like this:

```
foreach ($dataToValidate as $item) {
    $form = $this->formFactory->create($formType, $item, ['csrf_protection' => false, 'allow_extra_fields' => false]);
    $form->submit($item);

    if ($form->isSubmitted() && true === $form->isValid()) {
        $validData[] = $form->getData();
    }
}
```

My main issue is now, that although the form instance is no longer used, the memory is not released.

That looks like a bug to me.

### How to reproduce

To reproduce the behaviour you need run an iterative form validation (as in my example above).
In each iteration you will see that the memory usage increases and the used memory is never released.

**Comment by derrabus** [comment]
> I have an array of items (also as array) coming form a CSV file. Now I'm iterating over the items to validate the data using Symfony forms.

Are you aware that you can use the validator standalone? I'm really puzzled that you emulate hundrets of form submissions just to validate data.

> What I noticed now is that with every validation the memory usage goes up by around 0.29 MB and this memory never gets released. This for example leads to a memory increase from 20 to 70 MB after validating 200 items. (Each of the items is an array with 10 key / value pairs).

The form component isn't really designed to handle hundrets or thousands of for submits during a single request. I'm not saying that we shouldn't fix that, but your issue is certainly a rare edge-case.

Do you have an idea what's causing the memory leak?

**Comment by laurenz-glueck** [comment]
@derrabus Thanks for your response!
I already got the suggestion to use the validators standalone and also gave it a try. I guess that this is in general the better solution for my use case and it also seems not to have this memory leaks.

The problem is that we currently have about 120 form types in our project and converting them to just constrains will be quite a job...

I also have one more question you might can help me with:
When I have a from field of type `IntegerType` and in the request a number is provided as a string, the form submit also automatically converts the string to an int value. I guess this is done by "Data Transformers"? Is there also a way to archive this with standalone validators?

> Do you have an idea what's causing the memory leak?

Unfortunately not - I tried to debug it a bit and I noticed that the `$form->submit($item)` increases the memory. When you afterwords for example try to set the `$form` to `null` the memory is not getting less at all.

**Comment by derrabus** [comment]
> I also have one more question you might can help me with: When I have a from field of type `IntegerType` and in the request a number is provided as a string, the form submit also automatically converts the string to an int value. I guess this is done by "Data Transformers"? Is there also a way to archive this with standalone validators?

This is a hydration issue, not a validation issue. The validator does not modify the data it validates.

You can probably use the serializer here.

**Comment by stof** [comment]
Are you doing that in dev or in prod mode ? In dev mode, the profiler is hooked into form submission (for the `form` panel of the profiler), so the increased memory usage might be the collected data.

**Comment by laurenz-glueck** [comment]
@stof You are kind of right - I just changed my local setup to prod mode and the "issue" is gone.

I also tried to disable the profiler manually by injecting it into my class (https://symfony.com/doc/4.0/profiler/matchers.html) and then calling `$profiler->disable();` before running the form validation. But this somehow is not fixing my issue and the memory still gets filled up.

So I'm not sure if it is the profiler that is filling up the memory.
